File: ocr_reader.go
// ocr_reader.go — Go версия

package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/otiai10/gosseract/v2"
)

func main() {
	lang := flag.String("l", "eng", "Язык OCR (по умолчанию eng)")
	output := flag.String("o", "", "Файл для сохранения результата")
	flag.Parse()

	if flag.NArg() < 1 {
		fmt.Println("Usage: go run ocr_reader.go <image> [-l lang] [-o output]")
		os.Exit(1)
	}
	imagePath := flag.Arg(0)

	fmt.Println("\x1b[36m📖 OCR Reader (Go)\x1b[0m")
	fmt.Printf("📂 Обработка: %s\n", imagePath)
	fmt.Println("⏳ Распознавание...")

	start := time.Now()
	client := gosseract.NewClient()
	defer client.Close()
	client.SetLanguage(*lang)
	client.SetImage(imagePath)
	text, err := client.Text()
	elapsed := time.Since(start).Seconds()

	if err != nil {
		fmt.Printf("\x1b[31m❌ Ошибка: %v\x1b[0m\n", err)
		os.Exit(1)
	}

	text = strings.TrimSpace(text)
	fmt.Printf("\n\x1b[32mРезультат:\x1b[0m\n")
	fmt.Println(strings.Repeat("─", 50))
	fmt.Println(text)
	fmt.Println(strings.Repeat("─", 50))

	words := len(strings.Fields(text))
	chars := len(text)
	fmt.Printf("\x1b[36m📊 Статистика:\x1b[0m\n")
	fmt.Printf("  Слов: %d\n", words)
	fmt.Printf("  Символов: %d\n", chars)
	fmt.Printf("  Время: %.2f сек\n", elapsed)

	if *output != "" {
		os.WriteFile(*output, []byte(text), 0644)
		fmt.Printf("\x1b[32m💾 Сохранено в: %s\x1b[0m\n", *output)
	} else {
		outFile := imagePath[:len(imagePath)-len(extension(imagePath))] + ".txt"
		os.WriteFile(outFile, []byte(text), 0644)
		fmt.Printf("\x1b[32m💾 Сохранено в: %s\x1b[0m\n", outFile)
	}
}

func extension(path string) string {
	for i := len(path) - 1; i >= 0 && path[i] != '.'; i-- {
		if path[i] == '/' || path[i] == '\\' {
			return ""
		}
	}
	if i := strings.LastIndex(path, "."); i != -1 {
		return path[i:]
	}
	return ""
}

File: ocr_reader.php
<?php
// ocr_reader.php — PHP версия

require_once 'vendor/autoload.php';

use thiagoalessio\TesseractOCR\TesseractOCR;

$imagePath = null;
$lang = 'eng';
$outputFile = null;

$args = array_slice($argv, 1);
for ($i = 0; $i < count($args); $i++) {
    if ($args[$i] == '-l' || $args[$i] == '--lang') {
        $lang = $args[++$i];
    } elseif ($args[$i] == '-o' || $args[$i] == '--output') {
        $outputFile = $args[++$i];
    } elseif (!str_starts_with($args[$i], '-')) {
        $imagePath = $args[$i];
    }
}

if (!$imagePath) {
    echo "Usage: php ocr_reader.php <image> [-l lang] [-o output]\n";
    exit(1);
}

if (!file_exists($imagePath)) {
    echo "\033[31m❌ Файл не найден: $imagePath\033[0m\n";
    exit(1);
}

echo "\033[36m📖 OCR Reader (PHP)\033[0m\n";
echo "📂 Обработка: $imagePath\n";
echo "⏳ Распознавание...\n";

$start = microtime(true);
try {
    $ocr = new TesseractOCR($imagePath);
    $ocr->lang($lang);
    $text = trim((string) $ocr);
    $elapsed = microtime(true) - $start;

    echo "\n\033[32mРезультат:\033[0m\n";
    echo str_repeat("─", 50) . "\n";
    echo $text . "\n";
    echo str_repeat("─", 50) . "\n";

    $words = count(array_filter(explode(' ', $text)));
    $chars = strlen($text);
    echo "\033[36m📊 Статистика:\033[0m\n";
    echo "  Слов: $words\n";
    echo "  Символов: $chars\n";
    echo "  Время: " . number_format($elapsed, 2) . " сек\n";

    $outFile = $outputFile ?: pathinfo($imagePath, PATHINFO_FILENAME) . ".txt";
    file_put_contents($outFile, $text);
    echo "\033[32m💾 Сохранено в: $outFile\033[0m\n";
} catch (Exception $e) {
    echo "\033[31m❌ Ошибка: " . $e->getMessage() . "\033[0m\n";
    exit(1);
}
?>
